R/serialise.R

```diff
@@ -0,0 +1,79 @@
+##'
+##' * R has no scalar types so it is not clear if `1` should be
+##'   serialised as a number or a vector of length 1; jsonlite
+##'   provides support for "automatically unboxing" such values
+##'   (assuming that length-1 vectors are scalars) or never unboxing
+##'   them unless asked to using [jsonlite::unbox]
+##' * JSON has no date/time values and there are many possible string
+##'   representations.
+##' * JSON has no [data.frame] or [matrix] type and there are several
+##'   ways of representing these in JSON, all equally valid (e.g., row-wise,
+##'   column-wise or as an array of objects).
+##' * The handling of `NULL` and missing values (`NA`, `NaN`) are different
+##' * We need to chose the number of digits to write numbers out at,
+##'   balancing precision and storage.
+##'
+##' These issues are somewhat lessened when we have a schema because
+##' we know what our target type looks like.  This function attempts
+##' to use the schema to guide serialsation of json safely.  Currently
+##' it only supports detecting the appropriate treatment of length-1
+##' vectors, but we will expand functionality over time.
+##'
+##' For a user, this function provides an argument-free replacement
+##' for `jsonlite::toJSON`, accepting an R object and returning a
+##' string with the JSON representation of the object. Internally the
+##' algorithm is:
+##'
+##' 1. serialise the object with [jsonlite::toJSON], with
+##'    `auto_unbox = FALSE` so that length-1 vectors are serialised as a
+##'    length-1 arrays.
+##' 2. operating entirely within JavaScript, deserialise the object
+##'    with `JSON.parse`, traverse the object and its schema
+##'    simultaneously looking for length-1 arrays where the schema
+##'    says there should be scalar value and unboxing these, and
+##'    re-serialise with `JSON.stringify`
+##'
+##' There are several limitations to our current approach, and not all
+##' unboxable values will be found - at the moment we know that
+##' schemas contained within a `oneOf` block (or similar) will not be
+##' recursed into.
+##'
+##' @title Safe JSON serialisation
+##'
+##' @param x An object to be serialised
+##'
+##' @param schema A schema (string or path to a string, suitable to be
+##'   passed through to [jsonvalidate::json_validator] or a validator
+##'   object itself.
+##'
+##' @return A string, representing `x` in JSON format. As for
+##'   `jsonlite::toJSON` we set the class attribute to be `json` to
+##'   mark it as serialised json.
+##'
+##' @export
+##' @examples
+##' @example man-roxygen/example-json_serialise.R
+json_serialise <- function(x, schema) {
+  if (is.character(schema)) {
+    validator <- json_validator(schema, engine = "ajv")
+    v8 <- attr(validator, "v8", exact = TRUE)
+  } else if (inherits(schema, "function")) {
+    engine <- attr(schema, "engine", exact = TRUE)
+    if (engine != "ajv") {
+      stop("json_serialise is only supported with engine 'ajv'")
+    }
+    v8 <- attr(schema, "v8")
```

Comment:
Instead of pulling these things from `json_validator` you could copy the `json_validator` and `json_validate` pattern and add another function `json_serializer` analogous to `json_validator` which returns a function we can use to serialize.